### PR TITLE
`Integrated code lifecycle`: Simplify user interface for ssh keys

### PR DESCRIPTION
**Pull request by SimonEntholzer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The SSH Settings are not that easy to use for new users.

### Description

This PR improves the UI of the SSH Key settings.

### Steps for Testing

Prerequisites:
- Go to the SSH settings, and check that everything works as before.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

image
image
image
image
image

## Release Notes

- **New Features**
  - Introduced SSH key management capabilities, allowing users to view, add, and edit SSH keys.
  - Added a new property to enhance user information related to SSH keys in account details.
  - New component for generating documentation links dynamically.

- **Improvements**
  - Updated user interface for SSH settings to improve clarity and usability.
  - Enhanced documentation links for better guidance on SSH setup.
  - Streamlined rendering logic for SSH keys management.
  - Simplified string labels and added new entries for improved user instructions related to SSH key management.

- **Localization**
  - Updated and added strings for SSH key management in both English and German to improve user experience.

- **Bug Fixes**
  - Fixed minor issues related to the display and management of SSH keys in the user settings.

### CODE REVIEW

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    width: 100%;
+    border-collapse: collapse;
+    margin: 20px 0;
+    font-size: 16px;
+    text-align: left;
+}
+
+th,
+td {
+    padding: 15px 15px;
```
Comment (ole-ve) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss:21]:
```suggestion
    padding: 15px;
```

File: src/main/webapp/i18n/de/userSettings.json [hunk 2]
```diff
@@ -27,11 +27,22 @@
                 "deleteFailure": "SSH-Schlüssel konnte nicht gelöscht werden.",
                 "deleteSuccess": "SSH-Schlüssel erfolgreich gelöscht.",
                 "cancelSavingSshKey": "Abbrechen",
-                "editExistingSshKey": "SSH-Schlüssel bearbeiten",
-                "addNewSshKey": "Neuer SSH-Schlüssel",
+                "editExistingSshKey": "Bearbeiten",
+                "viewExistingSshKey": "Ansehen",
+                "addNewSshKey": "SSH Schlüssel hinzufügen",
+                "sshKeyDetails": "SSH Key Informationen",
```
Comment (ole-ve) [inline src/main/webapp/i18n/de/userSettings.json:33]: "Schlüssel" to be consistent?

File: src/main/webapp/i18n/en/userSettings.json [hunk 3]
```diff
@@ -27,11 +27,22 @@
...
+                "editExistingSshKey": "Edit",
+                "viewExistingSshKey": "View",
+                "addNewSshKey": "Add SSH key",
+                "sshKeyDetails": "SSH key details",
                 "deleteSshKey": "Delete",
                 "sshKeyDisplayedInformation": "This is your currently configured SSH key:",
-                "key": "SSH Key"
+                "key": "Key",
+                "noKeysHaveBeenAdded": "No SSH keys have been added",
+                "whatToUseSSHForInfo": "Use SSH keys to connect simply and safely to repositories.",
```
Comment (ole-ve) [inline src/main/webapp/i18n/en/userSettings.json:38]: securely?

**Review by ole-ve — COMMENTED** [review]
code looks good, added few smaller suggestions

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 4]
```diff
@@ -3,80 +3,153 @@ <h1 jhiTranslate="artemisApp.userSettings.sshSettings">
...
-                            (onClick)="saveSshKey()"
-                        />
+
+                <div class="d-flex flex-wrap mb-4">
+                    <p>
+                        <span style="font-size: small" jhiTranslate="artemisApp.userSettings.sshSettingsPage.alreadyHaveKey"> </span>
+                        <code> xclip -selection clipboard < ~/.ssh/id_ed25519.pub </code>
+                    </p>
+                </div>
+
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:119]: I agree with CodeRabbit here. This command is Linux only. A lot of Artemis users are using macOS or Windows, so providing instructions for those would be more helpful.

[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html]: I added a OS specific copy command

**Review by cremertim — DISMISSED** [review]
tested on ts2, works as described

**Review by BBesrour — APPROVED** [review]
Tested on TS2, works as expected

**Review by kaancayli — DISMISSED** [review]
Testted on ts2. Works as described 👍 

- Can see my existing ssh-key ✅ 
- Ssh-kep copy instruction is displayed correctly (MacOS) ✅ 
- Everything else works same as before ✅

File: src/main/webapp/content/scss/themes/_dark-variables.scss [hunk 5]
```diff
@@ -662,3 +662,9 @@ $iris-rate-background: var(--neutral-dark-l-15);
 
 // Image Cropper
 $cropper-overlay-color: transparent;
+
+// Settings
+$ssh-key-table-hover-background: $gray-900;
+$ssh-key-settings-dropdown-buttons: $gray-800;
+$ssh-key-settings-dropdown-buttons-hover: $gray-700;
+$ssh-key-settings-text-color: $white;
```
Comment (pzdr7) [inline src/main/webapp/content/scss/themes/_dark-variables.scss:670]: This variable is not present in the `_default-variables.scss`. Can you add it?

**Review by BBesrour — DISMISSED** [review]
code lgtm
and tested on ts2

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 6]
```diff
@@ -3,80 +3,153 @@ <h1 jhiTranslate="artemisApp.userSettings.sshSettings">
...
-            <div class="list-group-item">
-                {{ sshKey }}
-            </div>
-            <div class="list-group-item">
-                <div class="btn-group" role="group" aria-label="Actions">
+        <!-- Initial state: There are no keys -->
+        @if (keyCount === 0 && !showSshKey) {
+            <div class="list-group-item px-4 py-4">
+                <div class="mt-4"></div>
+                <h4 class="text-center mb-4 mt-8" jhiTranslate="artemisApp.userSettings.sshSettingsPage.noKeysHaveBeenAdded">No SSH keys have been added</h4>
```
Comment (pzdr7) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:10]: I think we agreed at one point to leave out the translation in the HTML

```suggestion
                <h4 class="text-center mb-4 mt-8" jhiTranslate="artemisApp.userSettings.sshSettingsPage.noKeysHaveBeenAdded"></h4>
```

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 7]
```diff
@@ -3,80 +3,153 @@ <h1 jhiTranslate="artemisApp.userSettings.sshSettings">
...
+                <div class="text-center mt-4 d-flex justify-content-center align-items-center flex-wrap" style="max-width: 500px; margin: 0 auto">
+                    <p>
+                        <span style="font-size: medium" jhiTranslate="artemisApp.userSettings.sshSettingsPage.whatToUseSSHForInfo"> </span>
```
Comment (pzdr7) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:13]: I would generally prefer css classes over inline styles with more than one element (like `max-width: 500px; margin: 0 auto`) or those you use multiple times (like `font-size: medium`)

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 8]
```diff
@@ -3,80 +3,153 @@ <h1 jhiTranslate="artemisApp.userSettings.sshSettings">
...
+                        <jhi-documentation-link [documentationType]="documentationType" [displayString]="'artemisApp.userSettings.sshSettingsPage.learnMore'">
+                        </jhi-documentation-link>
+                    </p>
+                </div>
+
+                <table class="table">
+                    <thead>
+                        <tr>
+                            <th jhiTranslate="artemisApp.userSettings.sshSettingsPage.keys">Keys</th>
+                            <th jhiTranslate="artemisApp.userSettings.sshSettingsPage.actions">Actions</th>
```
Comment (pzdr7) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:47]:
```suggestion
                            <th jhiTranslate="artemisApp.userSettings.sshSettingsPage.keys"></th>
                            <th jhiTranslate="artemisApp.userSettings.sshSettingsPage.actions"></th>
```

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html [hunk 9]
```diff
@@ -3,80 +3,153 @@ <h1 jhiTranslate="artemisApp.userSettings.sshSettings">
...
+                                            <button
+                                                jhiDeleteButton
+                                                class="btn dropdown-button"
+                                                [buttonSize]="ButtonSize.SMALL"
+                                                [renderButtonText]="false"
+                                                (delete)="deleteSshKey()"
+                                                deleteQuestion="artemisApp.userSettings.sshSettingsPage.deleteSshKeyQuestion"
+                                                [dialogError]="dialogError$"
+                                            >
+                                                <span jhiTranslate="artemisApp.userSettings.sshSettingsPage.deleteSshKey" class="ms-2">Delete</span>
```
Comment (pzdr7) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.html:79]:
```suggestion
                                                <span jhiTranslate="artemisApp.userSettings.sshSettingsPage.deleteSshKey" class="ms-2"></span>
```

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    transition: 0.35s ease-out;
+}
+
+/* dd content */
+.dropdown .dropdown-content {
+    position: absolute;
+    top: 50%;
+    min-width: 120%;
+    box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
+    z-index: 100000;
```
Comment (pzdr7) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss:74]: Maybe add a comment here explaining why we need to set the z-index to a high value

File: src/main/webapp/i18n/de/userSettings.json [hunk 11]
```diff
@@ -27,11 +27,22 @@
                 "deleteFailure": "SSH-Schlüssel konnte nicht gelöscht werden.",
                 "deleteSuccess": "SSH-Schlüssel erfolgreich gelöscht.",
                 "cancelSavingSshKey": "Abbrechen",
-                "editExistingSshKey": "SSH-Schlüssel bearbeiten",
-                "addNewSshKey": "Neuer SSH-Schlüssel",
+                "editExistingSshKey": "Bearbeiten",
+                "viewExistingSshKey": "Ansehen",
+                "addNewSshKey": "SSH Schlüssel hinzufügen",
+                "sshKeyDetails": "SSH Schlüssel Informationen",
```
Comment (pzdr7) [inline src/main/webapp/i18n/de/userSettings.json:33]: The other translations use a hyphen here:

```suggestion
                "addNewSshKey": "SSH-Schlüssel hinzufügen",
                "sshKeyDetails": "SSH-Schlüssel Informationen",
```

File: src/main/webapp/i18n/de/userSettings.json [hunk 12]
```diff
@@ -27,11 +27,22 @@
...
                 "deleteSshKey": "Löschen",
                 "sshKeyDisplayedInformation": "Das ist dein aktuell konfigurierter SSH-Schlüssel:",
-                "key": "SSH Schlüssel"
+                "key": "Schlüssel",
+                "noKeysHaveBeenAdded": "Es wurden noch keine SSH-Schlüssel hinzugefügt",
+                "whatToUseSSHForInfo": "Verwende SSH-Schlüssel, um einfach und sicher eine Verbindung zu Repositories herzustellen.",
+                "learnMore": "Erfahre mehr über SSH-Schlüssel",
+                "alreadyHaveKey": "Du hast bereits einen Schlüssel? Kopiere deinen Schlüssel in die Zwischenablage:",
+                "back": "Zurück",
+                "keysTablePageTitle": "SSH Schlüssel",
```
Comment (pzdr7) [inline src/main/webapp/i18n/de/userSettings.json:42]:
```suggestion
                "keysTablePageTitle": "SSH-Schlüssel",
```

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    cursor: pointer;
+    transition: 0.35s ease-out;
+}
+
+/* dd content */
+.dropdown .dropdown-content {
+    position: absolute;
+    top: 50%;
+    min-width: 120%;
+    box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
```
Comment (pzdr7) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss:73]: Could we define a color variable for this?

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Some suggestions on the code

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss (hunk 13 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss]: good suggestion :+1:

**Review by pzdr7 — DISMISSED** [review]
Code

File: src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+.dropdown {
+    display: inline-block;
+    position: relative;
+    outline: none;
+    margin: 0;
+}
+
+/* button */
+.dropbtn {
+    padding: 0 0;
```
Comment (kaancayli) [inline src/main/webapp/app/shared/user-settings/ssh-settings/ssh-user-settings.component.scss:71]:
```suggestion
    padding: 0;
```

**Review by kaancayli — APPROVED** [review]
Reapprove. One nitpick 😄

**Review by kaancayli — APPROVED** [review]

**Review by pzdr7 — APPROVED** [review]
Reapprove

**Review by BBesrour — APPROVED** [review]

**Pull request merged**